lib/src/models/binned.dart

```diff
@@ -550,53 +594,56 @@ abstract class BinnedChartState<T extends BinnedChart> extends State<T>
   /// and updates the selection controller if available.
   void _onTapUp(TapUpDetails details, AxisPainter axisPainter) {
     focusNode.requestFocus();
+    // Always remove the tooltip first
+    _clearHover();
     // Get the selected bin based on the tap location
     SelectedBin? selectedBin = _getBinOnTap(details.localPosition, axisPainter);
     _updatedBinSelection(selectedBin);
   }
 
   /// Update the selected bins based on the currently selected bin.
   void _updatedBinSelection(SelectedBin? selectedBin) {
-    selectedDataPoints = {};
-    if (selectedBin != null) {
-      if (selectedBins == null || selectedBins!.seriesIndex != selectedBin.seriesIndex || !isShifting) {
-        selectedBins = SelectedBinRange(selectedBin.seriesIndex, selectedBin.binIndex, null);
-      } else {
-        if (selectedBins!.startBinIndex == selectedBin.binIndex) {
-          selectedBins = null;
-        } else if (selectedBins!.startBinIndex < selectedBin.binIndex) {
-          selectedBins!.endBinIndex = selectedBin.binIndex;
-        } else {
-          selectedBins!.endBinIndex = selectedBins!.startBinIndex;
-          selectedBins!.startBinIndex = selectedBin.binIndex;
+    if (selectedBin == null) {
+      // Clear selection if no bin is selected
+      selectedBins?.clear();
+      _notifySelectionChange();
+      setState(() {});
+      return;
+    }
+
+    selectedBins ??= SelectedBins();
+
+    if (isShiftKeyPressed) {
+      // Range selection (Shift)
+      if (selectedBins!.selectedBins.containsKey(selectedBin.seriesIndex)) {
+        final Set<int> existingSelection = selectedBins!.selectedBins[selectedBin.seriesIndex]!;
+        int startIndex = existingSelection.reduce((a, b) => a < b ? a : b);
+        int endIndex = selectedBin.binIndex;
+
+        if (startIndex > endIndex) {
+          (startIndex, endIndex) = (endIndex, startIndex);
+        }
```

Comment:
I think that you might need to modify this slightly. I would get the startIndex as you currently do and set the initial endIndex as the last selected bin. Then check to see if `selectedBin.binIndex` is less than the `startIndex`, in which case set `startIndex = selectedBin.binIndex`, otherwise set `endIndex = selectedBin.binIndex`. I think that this behavior would be closer to what users expect from other apps.